### DISCUSSION TITLE
MidiPort.Send is obsolete since iOS 14 - no other way for sending MIDI via CoreMIDI on iOS available ?

### DISCUSSION THREAD

**Discussion by ayaphi** [dotnet/maui#12720] [Q&A]
Hi ithere,

I am using .NET MAUI, but focusing only on iOS right now, I cannot find a way (an Api) to send MIDI on iOS.

Naturally it would be CoreMidi and the `MidiPort` class using its `Send` method, but as the source states, since iOS 14 its not supported any more
```csharp       
[ObsoletedOSPlatform("macos11.0")]
[ObsoletedOSPlatform("ios14.0")]
[SupportedOSPlatform("ios")]
[SupportedOSPlatform("maccatalyst")]
[SupportedOSPlatform("macos")]
[UnsupportedOSPlatform("macos11.0")]
[UnsupportedOSPlatform("ios14.0")]
public MidiError Send(MidiEndpoint endpoint, MidiPacket[] packets);
```

I can retrieve a valid `MidiDevice`, `MidiEntity` and `MidiEndpoint` (which are related to the CoreMidi "Network" session / port), but after that I am stuck - theoretically I got all to call `MidiPort.Send` (see : https://learn.microsoft.com/en-us/dotnet/api/coremidi.midiport.send), but the method is not available in iOS 16.

And by the way, that is not the only obsolete CoreMIDI API - the Xamarin iOS MIDI sample has same issues : https://learn.microsoft.com/en-us/samples/xamarin/ios-samples/miditest/.

Any Ideas?

cheers 
Ayaphi

**Comment by ayaphi** [comment]
Hi @jfversluis,

as you have built a [audio plugin ](https://github.com/jfversluis/Plugin.Maui.Audio) for .NET MAUI, and thereby somehow into the musical sides of .NET MAUI ;-) and related to the team - what do you think about the outdated CoreMIDI API or even better do you have an idea for a solution? Are there plans for updating the API?

The API seems to be developed back in the days by Xamarin (see : https://tirania.org/monomac/archive/2012/Sep-11.html) and maybe updated only briefly and quite a long time ago.

I do not know how the iOS Plattform APIs are created for .NET MAUI, but chances are that this is done with a toolchain and a generative process - which makes it clear why the CoreMIDI API is outdated : its not a 1:1 projection, because back in the days Xamaring wrote a object oriented wrapper around the C alike API.

This in turn means : as long as no one will take the effort to update the OO wrapper classes the CoreMIDI API in .NET MAUI will stay outdated und practical useless for ages, because no automatic toolchain can pick it up und update it.

+++

I could imaginge a workaround, but this should be not the solution : there is open-source Swift framework MIDIKit (https://github.com/orchetect/MIDIKit) that wraps the objective-C based CoreMIDI API and seems to have a pretty nice OO design and this could be brought in by a "Binding Project".

(By the way : this MIDIKit library and its OO design could be a source of inspiration for a .NET MAUI team member to create a new or update the existing CoreMIDI OO Wrapper as well ;-)

But there are no .NET MAUI docs on binding native library and honestly I do not want to rely (trust ;-) the Xamarin iOS related docs on that (https://learn.microsoft.com/en-us/xamarin/ios/platform/binding-swift/), because the project system has changed significantly in .NET MAUI - this will end up in a lot of try and error (wasting time ;-).

I am aware of this .NET Conf talk (https://youtu.be/oibfI-ZsmzQ) and kindly the speaker has released a workflow (https://github.com/kinfey/AMapMAUIControls/blob/main/tutorial/en/01.iOSBinding.md) for binding native objective-c libraries, but which also show some problems of the toolchain and errors that have to fixed manually - which makes it obvious that this is still a work in progress. And there is no workflow for Swift Libraries.

cheers 
ayaphi

**Comment by xtuzy** [comment]
See source code of xamain's coremidi: https://github.com/xamarin/xamarin-macios/blob/10d4bb251eea94a742fd5891c0117867c0f6db57/src/CoreMidi/MidiServices.cs#L1108
xamatin's send method use MIDISend of Apple, see Apple doc:https://developer.apple.com/documentation/coremidi/1495289-midisend?language=objc, MIDISend is Deprecated, Apple say Use MIDISendEventList instead.

**Reply by ayaphi** [reply]
Thanks @xtuzy for pointing that out, but I see no way to call `MIDISendEventList` with the current .NET MAUI build / native iOS API bindings, because its nowhere accessible in the CoreMidi namesspace of .NET MAUI, or do I miss somthing?

But this is truly helpful for someone who updates the exisintg OO wrapper in .NET MAUI carried over from Xamarin.

But this is not the only API related to Audio & Midi that is missing or outdated - for example the `AVFAudio` framework, which is super usefull with its `AVAudioEngine` (you build a graph of nodes in there) and `AVAudioSequencer` classes for **professional grade audio inside an app** (theses classes cannot however deliver MIDI outside the app though, that's what CoreMIDI framework is needed for), including playing back ones own midi with soundfonts e.g (midi files or midi events "sequenced" at runtime) .

`AVMusicTrack` (https://developer.apple.com/documentation/avfaudio/avmusictrack), which is needed e.g. for adding and removing MIDI events to a established sequencer session, **misses ALL methods in .NET MAUI** : https://learn.microsoft.com/en-us/dotnet/api/avfoundation.avmusictrack (and be aware of the URL redirect when you follow this link - this hints somehow to the state of API binding updates, at least documention wise, but somehow also "real code" wise : we are stuck on older versions).

It seems, that the team is already aware of this, see : https://github.com/xamarin/xamarin-macios/issues/4452, as @rolfbjarne has attached the label "missing-api-bindings".

I guess the entire Audio & MIDI native iOS API bindings needs a thoughtful overhaul, and not only a run of an automated toolchain to get the API bindings updated (if so, they would have done it already ;-), as e.g. the CoreMIDI API has an OO wrapper in .NET MAUI, which involves a "manuall implementation" and thoughtful OO design process in order to be "future proof" on iOS.

That's because the strategy of Apple at least regarding "audio across / between apps" has changed over the time and if .NET MAUI still wants to provide OO wrappers in that field, they should reflect the "future strategy" : as far as I know "inter-app-audio" has been annouced to be deprecated in some time in the future and that AUv3 is the way to go. 

Disclaimer : I know this only from a consumer perspective right now and have not digged into the orginal announcement of Apple yet, but I've read this several times in the audiobus forum (https://forum.audiob.us), which seems to be the epicenter for iOS music apps & iOS music production : audio & midi wise).

Regaring MIDI : I am not aware of annoucements regarding MIDI, but I think CoreMIDI won't be deprecated as a whole (in the sense of a technology or framework), as there has to be a way to input MIDI from outside iOS devices (foremost hardware MIDI keyboards & controllers, but also other source via network) into the ecosystem of music apps - that's the job of CoreMIDI and chances are, it will be there in the future as well - it's too fundamental and as far as I know, Apple has update the API to the MIDI 2.0 specs, which signals that CoreMIDI is really kept up to date.

cheers
Ayaphi

**Reply by xtuzy** [reply]
i'm not developer of microsoft, you say these to me not help you solve problem. MIDISendEventList method need MIDIEventList, i see have a issue about it :https://github.com/xamarin/xamarin-macios/issues/12489 , you can try to "help" xamarin.ios team to binding it (tips: they have a discord channel, you can ask how to binding it)

**Reply by ayaphi** [reply]
Hi @xtuzy,

please don't take my message personally - what I wrote was by no means meant to be an advice or "call to action" or whatsoever for you (or to you) to do something nor to be responsible for something.

Please, take it easy!

For me it was just a way to submit information, not only to you, but also to all the other users that might read this thread. As you brought up "a thing" I just mentioned and adressed you, so that you get informed, also give respect to your findings and maybe giving you also the "chance" (can't find the right words here) to benefit from beeing ware of what I have found out.

Just to be clear : even if this post was headed with "Hi tuzy", it also and even formost also includes and meant (even if it was not stated) "& Hi to everyone who might read this", which might helpfully also be someone of the .NET MAUI team someday.

Each post, each message not only goes out to the one who is mentioned or greeted or to whose comment it is attached to (as a following comment), it also and foremost goes out to the entire community in the hope that someone who has the time or the resposibility will read this and be inspired to "implement a change".

Again thank for your hints to the issue and that Xamarin has a Discord channel - did not know it.

And please don't feel upset - this was by no means my intenion.

Have a nice day & cheers
Ayaphi
